### ISSUE TITLE
Allow fine-grained test management with 'test sets'

### ISSUE THREAD

**Issue by ellenkorbes** [garden-io/garden#851]
As originally mentioned in #438, users might not want to run _all_ tests _all_ the time.

I suggest we allow the user to define named sets of tests to run on different scenarios. Some ideas:

- A `default` test set defines which tests run on every file change, when `garden dev` is running.
- If `default` is undefined, it contains all tests.
- An `all` test set is hard-coded to include all tests.
- A `--testset` flag allows the user to manually specify which test set to use.

If we're feeling fancy, we could allow test sets to be test sets, which should give more granular control—but could also be completely useless. 

For example, let's say I have three tests, `t1`, `t2`, and `t3`. I define a test set called `mytestset` containing those three tests. I can then specify `default: mytestset`.

OR I have three tests, `t1`, `t2`, and `t3`. I define a test set called `mytestset` containing tests `t1` and `t2`. I can then specify `default: mytestset, t3`.

Regardless of these suggestions, _any_ way to differentiate between tests to run on `garden dev` and tests to run in CI would solve the problem.

**Comment by edvald** [comment]
So, we already have names on tests. And `garden test` allows you to filter by those, but not currently `garden dev`. Adding a `--test-name` flag to the dev command would be pretty easy.

Then there's the notion of test groups/sets, which could be quite nice. Basically we could have a `group` field on test specs, which defaults to "default", and then `--group/--test-group` field on the test+dev commands respectively.

How does that sound?

**Comment by ellenkorbes** [comment]
Sounds good.

Do we prefer to list the groups a test belongs to when defining the tests, or would we rather have a groups field and list the tests there per group?

```
tests:
  - groups:
    - myGroup: test1, test2, test3
  - name: test1
    ...
  - name: test2
    ...
  - name: test3
    ...
```

or

```
tests:
  - name: test1
    groups: myGroup
    ...
  - name: test2
    groups: myGroup
    ...
  - name: test3
    groups: myGroup
    ...
```

?

**Comment by 10ko** [comment]
So as far as I am concerned the list of things we should implement/update is the following:

 - A `--skip-tests` flag to skip all the tests in `garden dev`.
 - Introduction of concept of test-groups. 
   - Group definition. I very much prefer the former approach of "*creating a groups field and list the tests there per group*" @ellenkorbes proposed. Group names are not unique.
   - `--test-groups` flag to select the groups.
 - To make things more homogenous and less verbose in terms of configuration, I will add also the option of specifying test names like in the `garden dev` command (`-n, --name` flag). This is in case the user doesn't want to specify a test group.
- A `--test-pattern` (?) flag matching only certain test file names? (is that something we would want)? Please let me know what you think.

And I suggest we take it from here and see how good it feels. If we need more or less control or a different way of defining groups we can iterate on this.

I could also imagine to define the groups in a `.garden-test.yml` file with even more granular control but seems overkill right now.

Please let me know if you have any concerns, question or if I am missing something.
Peace,
E.

**Comment by edvald** [comment]
I remain sceptical of the notion of groups, since it looks to me like the key use cases can be satisfied with names. For example, we generally use the names "unit" and "integ" across our example projects. I can't quite think of a pressing use case for an additional field/structure, and I'd rather like to take more time to consider the notion of groups for other entities as well. Unless I'm missing something?

Re: the name and pattern flags, my suggestion would be to allow glob-style wildcards in the `--name/--test-name` field (for `garden dev` it should be `--test-name` because otherwise it's ambiguous). That way an additional `--pattern` field should be mostly unnecessary.

**Comment by 10ko** [comment]
Implemented in #917:
 - `garden dev --skip-tests`
 - glob-style wildcards for `garden test --name integ*`
 - glob-style wildcards and comma separated test name list for `garden dev --test-names integ*,unit`

Closing.
